torch/utils/checkpoint.py

```diff
@@ -364,78 +660,55 @@ def _checkpoint_without_reentrant(function, preserve_rng_state=True, *args, **kw
-                if preserve_rng_state:
-                    torch.set_rng_state(fwd_cpu_state)
-                    if had_cuda_in_fwd:
-                        set_device_states(fwd_gpu_devices, fwd_gpu_states)
-
-                with torch.enable_grad(), \
-                     torch.cuda.amp.autocast(**gpu_autocast_kwargs), \
-                     torch.cpu.amp.autocast(**cpu_autocast_kwargs), \
-                     torch.autograd.graph.saved_tensors_hooks(inner_pack, inner_unpack):
-                    _unused = function(*args, **kwargs)
-
-        if x not in storage:
-            raise RuntimeError(
-                "Attempt to retrieve a tensor saved by autograd multiple times without checkpoint"
-                " recomputation being triggered in between, this is not currently supported. Please"
-                " open an issue with details on your use case so that we can prioritize adding this."
-            )
-
-        return storage[x]
+    flat_args, kwarg_keys = _pack_kwargs(*args, **kwargs)
 
-    with torch.autograd.graph.saved_tensors_hooks(pack, unpack):
-        output = function(*args, **kwargs)
+    def new_fn(*inputs):
+        # This function should be called immediately by checkpoint_impl
+        unpacked_args, unpacked_kwargs = _unpack_kwargs(
+            inputs, kwarg_keys
+        )
+        out = fn(*unpacked_args, **unpacked_kwargs)
         if torch.cuda._initialized and preserve_rng_state and not had_cuda_in_fwd:
             # Cuda was not initialized before running the forward, so we didn't
             # stash the CUDA state.
             raise RuntimeError(
                 "PyTorch's CUDA state was initialized in the forward pass "
                 "of a Checkpoint, which is not allowed. Please open an issue "
                 "if you need this feature.")
+        return out
+
+    def recompute_fn(*inputs):
+        # This will be called later during recomputation. This wrapping enables
+        # the necessary global state to be captured.
+        unpacked_args, unpacked_kwargs = _unpack_kwargs(
+            inputs, kwarg_keys
+        )
+
+        rng_devices = []
+        if preserve_rng_state and had_cuda_in_fwd:
+            rng_devices = fwd_gpu_devices
+        with torch.random.fork_rng(devices=rng_devices, enabled=preserve_rng_state):
+            if preserve_rng_state:
+                torch.set_rng_state(fwd_cpu_state)
+                if had_cuda_in_fwd:
+                    set_device_states(fwd_gpu_devices, fwd_gpu_states)
+
+            with torch.cuda.amp.autocast(**gpu_autocast_kwargs), \
+                 torch.cpu.amp.autocast(**cpu_autocast_kwargs):
+                fn(*unpacked_args, **unpacked_kwargs)
+
+    return _checkpoint_impl(new_fn, recompute_fn, *flat_args)
+
+def _checkpoint_impl(fn, recompute_fn, *args):
+    new_frame = _CheckpointFrame(recompute_fn)
+
+    dummy = torch.empty((0,), requires_grad=True)
+    new_frame.input_saver = _NoopSaveInputs.apply(dummy, *args)
```

Comment:
Yeah, this doesn't require unpacked args. Just pass them as-is.

Reply:
I tried rewriting it this way, but I feel like its easier later on if we unpack here - I'd like to have(dummy, *args, kwargs) all squished together in a single tuple so that I can just do `tensor_indices, tensors = zip(*[(i, o) for i, o in enumerate(inputs) if isinstance(o, torch.Tensor)])` in one go later. Notably I need to include dummy as well, because there are cases where there are no tensors that are not hidden in data structures.